File: LeetCodeSolutions/LeetCode_0034.py
from typing import List


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        # Bineary Search
        if not nums:
            return [-1, -1]

        left, right = 0, len(nums)
        while left < right:
            mid = (left + right) // 2
            if nums[mid] == target:
                right = mid
            elif nums[mid] > target:
                right = mid
            elif nums[mid] < target:
                left = mid + 1
        if left == len(nums):
            return [-1, -1]
        if nums[left] == target:
            idx1 = left
        else:
            return [-1, -1]
        left, right = 0, len(nums)
        while left < right:
            mid = (left + right) // 2
            if nums[mid] == target:
                left = mid + 1
            elif nums[mid] > target:
                right = mid
            elif nums[mid] < target:
                left = mid + 1
        return [idx1, left - 1]

File: LeetCodeSolutions/LeetCode_0034.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int l = 0, r = nums.size();
        while (l < r){
            int mid = l + (r - l) / 2;
            if (nums[mid] < target) l = mid + 1;
            else r = mid;
        }
        if (l == nums.size()) return {-1, -1};
        if (nums[l] != target) return {-1, -1};
        int a = 0, b = nums.size();
        while (a < b){
            int m  = a + (b - a) / 2;
            if (nums[m] <= target) a = m + 1;
            else b = m;
        }
        return {l, a - 1};
    }
};